File: src/controllers/port-mapper.ts
import { UnifiController } from './unifi';
import { NodePortMapping } from '../types';
import logger from '../utils/logger';

interface CachedData {
  timestamp: number;
  data: any;
}

export class PortMapperController {
  private unifiController: UnifiController;
  private cache: Map<string, CachedData> = new Map();
  private readonly CACHE_TTL = 60000; // 1 minute cache
  private readonly QUERY_TIMEOUT = 3000; // 3 second timeout per query

  constructor(unifiController: UnifiController) {
    this.unifiController = unifiController;
  }

  private isCacheValid(key: string): boolean {
    const cached = this.cache.get(key);
    if (!cached) return false;
    return Date.now() - cached.timestamp < this.CACHE_TTL;
  }

  private getCached<T>(key: string): T | null {
    if (this.isCacheValid(key)) {
      return this.cache.get(key)!.data as T;
    }
    return null;
  }

  private setCache(key: string, data: any): void {
    this.cache.set(key, {
      timestamp: Date.now(),
      data,
    });
  }

  private async withTimeout<T>(promise: Promise<T>, timeoutMs: number): Promise<T> {
    return Promise.race([
      promise,
      new Promise<T>((_, reject) =>
        setTimeout(() => reject(new Error('Timeout')), timeoutMs)
      ),
    ]);
  }

  async getNodePortMappings(nodes: { name: string; ip?: string; mac?: string }[]): Promise<NodePortMapping[]> {
    logger.info('Getting port mappings for all nodes...');
    const startTime = Date.now();

    try {
      // Check if we have cached data for all nodes
      const cacheKey = 'all-nodes-mappings';
      const cached = this.getCached<NodePortMapping[]>(cacheKey);
      if (cached) {
        logger.info(`Using cached port mappings (${Date.now() - startTime}ms)`);
        return cached;
      }

      // Fetch all necessary data in parallel with timeouts
      const [devices, clients] = await Promise.all([
        this.withTimeout(this.unifiController.getDevices(), this.QUERY_TIMEOUT)
          .catch(err => {
            logger.warn('Failed to fetch devices:', err.message);
            return [];
          }),
        this.withTimeout(this.unifiController.getAllClients(), this.QUERY_TIMEOUT)
          .catch(err => {
            logger.warn('Failed to fetch clients:', err.message);
            return [];
          }),
      ]);

      // Filter for switches only
      const switches = devices.filter(d => d.type === 'usw' || d.type === 'usg');
      
      if (switches.length === 0) {
        logger.warn('No switches found');
        return [];
      }

      // Fetch all switch port tables in parallel with timeouts
      const switchPortPromises = switches.map(async (switchDevice) => {
        try {
          const ports = await this.withTimeout(
            this.unifiController.getSwitchPorts(switchDevice.mac),
            this.QUERY_TIMEOUT
          );
          return {
            switchDevice,
            ports,
          };
        } catch (err: any) {
          logger.warn(`Failed to fetch ports for switch ${switchDevice.name}:`, err.message);
          return {
            switchDevice,
            ports: [],
          };
        }
      });

      const switchPortResults = await Promise.all(switchPortPromises);

      // Build mapping for each node
      const mappings: NodePortMapping[] = nodes.map(node => {
        const nodeMac = node.mac?.toLowerCase();
        const nodeIp = node.ip;
        
        // Find client by MAC (preferred) or IP address (fallback)
        let client = nodeMac ? clients.find(c => c.mac?.toLowerCase() === nodeMac) : null;
        
        // If not found by MAC, try matching by IP
        if (!client && nodeIp) {
          client = clients.find(c => c.ip === nodeIp);
          if (client) {
            logger.info(`Matched node ${node.name} by IP address ${nodeIp}`);
          }
        } else if (client) {
          logger.info(`Matched node ${node.name} by MAC address ${nodeMac}`);
        }
        
        if (!client || !client.sw_mac) {
          logger.warn(`Unable to find UniFi client for node ${node.name} (IP: ${nodeIp}, MAC: ${nodeMac || 'N/A'})`);
          return {
            nodeName: node.name,
            switchName: 'Not Connected',
            portIdx: 0,
            poeAvailable: false,
          };
        }

        // Find the switch this client is connected to
        const switchInfo = switchPortResults.find(
          s => s.switchDevice.mac.toLowerCase() === client.sw_mac.toLowerCase()
        );

        if (!switchInfo) {
          return {
            nodeName: node.name,
            switchName: 'Switch Not Found',
            portIdx: client.sw_port || 0,
            poeAvailable: false,
          };
        }

        // Find the port
        const port = switchInfo.ports.find(p => p.port_idx === client.sw_port);
        
        return {
          nodeName: node.name,
          switchName: switchInfo.switchDevice.name || switchInfo.switchDevice.mac,
          switchMac: switchInfo.switchDevice.mac,
          portIdx: client.sw_port || 0,
          poeAvailable: port ? (port.port_poe && port.poe_enable) : false,
        };
      });

      // Cache the results
      this.setCache(cacheKey, mappings);

      const elapsed = Date.now() - startTime;
      logger.info(`Port mappings retrieved successfully in ${elapsed}ms`);
      return mappings;

    } catch (error: any) {
      const elapsed = Date.now() - startTime;
      logger.error(`Failed to get port mappings after ${elapsed}ms:`, error.message);
      
      // Return empty mappings for all nodes
      return nodes.map(node => ({
        nodeName: node.name,
        switchName: 'Error',
        portIdx: 0,
        poeAvailable: false,
      }));
    }
  }

  async powerCycleNodePort(nodeName: string, nodes: { name: string; ip?: string; mac?: string }[]): Promise<void> {
    logger.info(`Power cycling port for node ${nodeName}...`);

    try {
      // Get the port mapping for this specific node
      const mappings = await this.getNodePortMappings(nodes);
      const mapping = mappings.find(m => m.nodeName === nodeName);

      if (!mapping || !mapping.switchMac) {
        throw new Error('Port mapping not found for node');
      }

      if (!mapping.poeAvailable) {
        throw new Error('Node is not connected to a PoE-enabled port');
      }

      await this.unifiController.powerCyclePort(mapping.switchMac, mapping.portIdx);
      
      // Invalidate cache after power cycle
      this.cache.delete('all-nodes-mappings');
      
      logger.info(`Successfully power cycled port for node ${nodeName}`);
    } catch (error: any) {
      logger.error(`Failed to power cycle port for node ${nodeName}:`, error.message);
      throw error;
    }
  }

  clearCache(): void {
    logger.info('Clearing port mapper cache');
    this.cache.clear();
  }
}

File: src-tauri/src/controllers/port_mapper.rs
use anyhow::{bail, Context, Result};
use std::time::{Duration, Instant};
use tokio::sync::Mutex;
use tracing::{info, warn};

use crate::types::{K8sNode, NodePortMapping};
use crate::controllers::unifi::UnifiClient;
use crate::state::PortMappingCache;

const CACHE_TTL: Duration = Duration::from_secs(60);

pub async fn get_node_port_mappings(
    nodes: &[K8sNode],
    unifi: &Mutex<UnifiClient>,
    cache: &Mutex<Option<PortMappingCache>>,
) -> Vec<NodePortMapping> {
    // Return cached result if still fresh
    {
        let guard = cache.lock().await;
        if let Some(c) = guard.as_ref() {
            if c.timestamp.elapsed() < CACHE_TTL {
                info!("Using cached port mappings");
                return c.data.clone();
            }
        }
    }

    info!("Fetching fresh port mappings...");

    let (clients, switch_ports) = tokio::join!(
        async {
            let mut u = unifi.lock().await;
            u.get_all_clients().await.unwrap_or_else(|e| {
                warn!("Failed to fetch UniFi clients: {}", e);
                vec![]
            })
        },
        async {
            let mut u = unifi.lock().await;
            u.get_all_switch_ports().await.unwrap_or_else(|e| {
                warn!("Failed to fetch switch ports: {}", e);
                vec![]
            })
        }
    );

    let mappings: Vec<NodePortMapping> = nodes
        .iter()
        .map(|node| {
            // Try matching client by MAC, then by IP
            let client = node
                .ip
                .as_deref()
                .and_then(|ip| {
                    clients.iter().find(|c| {
                        c.get("ip").and_then(|v| v.as_str()) == Some(ip)
                    })
                })
                .or_else(|| {
                    match &node.mac {
                        Some(mac) => {
                            let mac_lower = mac.to_lowercase();
                            clients.iter().find(|c| {
                                c.get("mac")
                                    .and_then(|v| v.as_str())
                                    .map(|m| m.to_lowercase() == mac_lower)
                                    .unwrap_or(false)
                            })
                        }
                        None => None,
                    }
                });

            let Some(client) = client else {
                warn!("No UniFi client found for node {}", node.name);
                return NodePortMapping {
                    node_name: node.name.clone(),
                    switch_name: "Not Connected".into(),
                    switch_mac: None,
                    port_idx: 0,
                    poe_available: false,
                };
            };

            let sw_mac = client.get("sw_mac").and_then(|v| v.as_str()).unwrap_or("");
            let sw_port = client.get("sw_port").and_then(|v| v.as_u64()).unwrap_or(0) as u32;

            let sw_mac_lower = sw_mac.to_lowercase();
            let switch_info = switch_ports.iter().find(|(mac, _, _)| mac.to_lowercase() == sw_mac_lower);

            let Some((switch_mac, switch_name, ports)) = switch_info else {
                return NodePortMapping {
                    node_name: node.name.clone(),
                    switch_name: "Switch Not Found".into(),
                    switch_mac: None,
                    port_idx: sw_port,
                    poe_available: false,
                };
            };

            let port = ports.iter().find(|p| p.port_idx == Some(sw_port));
            let poe_available = port
                .map(|p| p.port_poe.unwrap_or(false) && p.poe_enable.unwrap_or(false))
                .unwrap_or(false);

            NodePortMapping {
                node_name: node.name.clone(),
                switch_name: switch_name.clone(),
                switch_mac: Some(switch_mac.clone()),
                port_idx: sw_port,
                poe_available,
            }
        })
        .collect();

    // Update cache
    {
        let mut guard = cache.lock().await;
        *guard = Some(PortMappingCache {
            timestamp: Instant::now(),
            data: mappings.clone(),
        });
    }

    mappings
}

pub async fn power_cycle_node_port(
    node_name: &str,
    nodes: &[K8sNode],
    unifi: &Mutex<UnifiClient>,
    cache: &Mutex<Option<PortMappingCache>>,
) -> Result<()> {
    let mappings = get_node_port_mappings(nodes, unifi, cache).await;
    let mapping = mappings
        .iter()
        .find(|m| m.node_name == node_name)
        .with_context(|| format!("Port mapping not found for node {}", node_name))?;

    if !mapping.poe_available {
        bail!("Node {} is not connected to a PoE-enabled port", node_name);
    }

    let switch_mac = mapping
        .switch_mac
        .as_deref()
        .context("Switch MAC not available")?;

    let mut unifi_guard = unifi.lock().await;
    unifi_guard.power_cycle_port(switch_mac, mapping.port_idx).await?;

    // Invalidate cache after power cycle
    {
        let mut guard = cache.lock().await;
        *guard = None;
    }

    Ok(())
}
